conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+    first dimension of the batch).  
+    
+    Unlike nn.DataParallel, all of the wrapped module's parameters will be 
+    copied to all gpus during construction of the wrapper, and only gradient 
+    data will be copied from gpu to gpu during training (no data should need 
+    to be copied between gpus during forward application).  However, training 
+    should only use loss functions that can be expressed as averages over all 
+    data points of a batch of some per-data-point loss.  Also, the set of 
+    parameters of the wrapped module is assumed not to change (of course their 
+    values can change) over the lifetime of the wrapper object.
+    
+    Note that during training of a wrapped module, it is necessary to call 
+    the wrapper's broadcast_gradients method immediately following the 
+    backpropagation of gradients (i.e. typically right after calling 
+    .backward on some loss tensor), in order to share corrections to the 
+    computed gradients between the gpus.
+    """
+    def __init__(self, module, device, copy_fn=copy.deepcopy):
+        """
+        Construct a parallelizing wrapper for the given module (an instance 
+        of nn.Module).  The module will be moved to the given device (if 
+        not already there) and copies will be made using the given copy 
+        function (defaulting to copy.deepcopy) and placed on all other 
+        available gpus.
+        """
+        super().__init__()
+        device_ids = list(range(torch.cuda.device_count()))
+        devices = [torch.device('cuda:{}'.format(d_id)) for d_id in device_ids]
+        if device not in devices:
+            device = devices[0]
+        self.children = [module]
+        self.devices = [device]
+        self.add_module('child:{}'.format(device.index), module)
+        for dev in devices:
+            if dev != device:
+                self.devices.append(dev)
+        for dev in self.devices[1:]:
+            module.cpu()
+            module_copy = copy_fn(module)
+            module_copy.to(dev)
+            self.children.append(module_copy)
+            self.add_module('child:{}'.format(dev.index), module_copy)
+        module.to(device)
+        self.chunk_sizes = torch.zeros(len(self.children), dtype=torch.int64)
+
+    def forward(self, *args):
+        """
+        Scatter the supplied args (assumed to be a list of tensors) across 
+        the child modules, and gather their outputs (assumed to be single 
+        tensors) back to the first gpu.  Also, accumulate the sizes of the 
+        scattered chunks (for later use in updating parameter gradients).
+        """
+        if len(self.children) <= 1:
+            return self.children[0](*args)
+        device_ids = [device.index for device in self.devices]
+        chunk_lists = list(list() for i_device in device_ids)
+        for arg in args:
+            chunks = torch.cuda.comm.scatter(arg, device_ids)
```

Comment:
What does it specifically mean to "scatter"?

Reply:
I think it is a term of art in the parallel computing world; scattering some data item means breaking it into shards that are sent to different processors for their consumption. The docs for torch.cuda.comm.scatter are at https://pytorch.org/docs/stable/cuda.html; it creates a tuple of shards of its input tensor, each shard on one of the devices specified by id.